// zwave-js/build/lib/node/mockCCBehaviors/ManufacturerSpecific.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManufacturerSpecificCCBehaviors = void 0;
const ManufacturerSpecificCC_1 = require("@zwave-js/cc/ManufacturerSpecificCC");
const testing_1 = require("@zwave-js/testing");
const respondToManufacturerSpecificGet = {
    async onControllerFrame(controller, self, frame) {
        if (frame.type === testing_1.MockZWaveFrameType.Request
            && frame.payload instanceof ManufacturerSpecificCC_1.ManufacturerSpecificCCGet) {
            const cc = new ManufacturerSpecificCC_1.ManufacturerSpecificCCReport(self.host, {
                nodeId: self.id,
                manufacturerId: self.capabilities.manufacturerId,
                productType: self.capabilities.productType,
                productId: self.capabilities.productId,
            });
            await self.sendToController((0, testing_1.createMockZWaveRequestFrame)(cc, {
                ackRequested: false,
            }));
            return true;
        }
    },
};
exports.ManufacturerSpecificCCBehaviors = [
    respondToManufacturerSpecificGet,
];
