// zwave-js/build/lib/serialapi/application/ApplicationUpdateRequest.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationUpdateRequestSmartStartHomeIDReceived = exports.ApplicationUpdateRequestNodeRemoved = exports.ApplicationUpdateRequestNodeAdded = exports.ApplicationUpdateRequestNodeInfoRequestFailed = exports.ApplicationUpdateRequestNodeInfoReceived = exports.ApplicationUpdateRequestWithNodeInfo = exports.ApplicationUpdateRequest = exports.ApplicationUpdateTypes = void 0;
const core_1 = require("@zwave-js/core");
const serial_1 = require("@zwave-js/serial");
const shared_1 = require("@zwave-js/shared");
var ApplicationUpdateTypes;
(function (ApplicationUpdateTypes) {
    ApplicationUpdateTypes[ApplicationUpdateTypes["SmartStart_NodeInfo_Received"] = 134] = "SmartStart_NodeInfo_Received";
    ApplicationUpdateTypes[ApplicationUpdateTypes["SmartStart_HomeId_Received"] = 133] = "SmartStart_HomeId_Received";
    ApplicationUpdateTypes[ApplicationUpdateTypes["NodeInfo_Received"] = 132] = "NodeInfo_Received";
    ApplicationUpdateTypes[ApplicationUpdateTypes["NodeInfo_RequestDone"] = 130] = "NodeInfo_RequestDone";
    ApplicationUpdateTypes[ApplicationUpdateTypes["NodeInfo_RequestFailed"] = 129] = "NodeInfo_RequestFailed";
    ApplicationUpdateTypes[ApplicationUpdateTypes["RoutingPending"] = 128] = "RoutingPending";
    ApplicationUpdateTypes[ApplicationUpdateTypes["Node_Added"] = 64] = "Node_Added";
    ApplicationUpdateTypes[ApplicationUpdateTypes["Node_Removed"] = 32] = "Node_Removed";
    ApplicationUpdateTypes[ApplicationUpdateTypes["SUC_IdChanged"] = 16] = "SUC_IdChanged";
})(ApplicationUpdateTypes || (exports.ApplicationUpdateTypes = ApplicationUpdateTypes = {}));
const { decorator: applicationUpdateType, lookupConstructor: getApplicationUpdateRequestConstructor, lookupValue: getApplicationUpdateType, } = (0, core_1.createSimpleReflectionDecorator)({
    name: "applicationUpdateType",
});
let ApplicationUpdateRequest = class ApplicationUpdateRequest extends serial_1.Message {
    constructor(host, options) {
        super(host, options);
        if ((0, serial_1.gotDeserializationOptions)(options)) {
            this.updateType = this.payload[0];
            const CommandConstructor = getApplicationUpdateRequestConstructor(this.updateType);
            if (CommandConstructor
                && new.target !== CommandConstructor) {
                return new CommandConstructor(host, options);
            }
            this.payload = this.payload.subarray(1);
        }
        else {
            this.updateType = getApplicationUpdateType(this);
        }
    }
    updateType;
    serialize() {
        this.payload = Buffer.concat([
            Buffer.from([this.updateType]),
            this.payload,
        ]);
        return super.serialize();
    }
};
exports.ApplicationUpdateRequest = ApplicationUpdateRequest;
exports.ApplicationUpdateRequest = ApplicationUpdateRequest = __decorate([
    (0, serial_1.messageTypes)(serial_1.MessageType.Request, serial_1.FunctionType.ApplicationUpdateRequest)
    // this is only received, not sent!
], ApplicationUpdateRequest);
class ApplicationUpdateRequestWithNodeInfo extends ApplicationUpdateRequest {
    constructor(host, options) {
        super(host, options);
        if ((0, serial_1.gotDeserializationOptions)(options)) {
            this.nodeInformation = (0, core_1.parseNodeUpdatePayload)(this.payload, this.host.nodeIdType);
            this.nodeId = this.nodeInformation.nodeId;
        }
        else {
            this.nodeId = options.nodeInformation.nodeId;
            this.nodeInformation = options.nodeInformation;
        }
    }
    nodeId;
    nodeInformation;
    serialize() {
        this.payload = (0, core_1.encodeNodeUpdatePayload)(this.nodeInformation, this.host.nodeIdType);
        return super.serialize();
    }
}
exports.ApplicationUpdateRequestWithNodeInfo = ApplicationUpdateRequestWithNodeInfo;
let ApplicationUpdateRequestNodeInfoReceived = class ApplicationUpdateRequestNodeInfoReceived extends ApplicationUpdateRequestWithNodeInfo {
};
exports.ApplicationUpdateRequestNodeInfoReceived = ApplicationUpdateRequestNodeInfoReceived;
exports.ApplicationUpdateRequestNodeInfoReceived = ApplicationUpdateRequestNodeInfoReceived = __decorate([
    applicationUpdateType(ApplicationUpdateTypes.NodeInfo_Received)
], ApplicationUpdateRequestNodeInfoReceived);
let ApplicationUpdateRequestNodeInfoRequestFailed = class ApplicationUpdateRequestNodeInfoRequestFailed extends ApplicationUpdateRequest {
    isOK() {
        return false;
    }
};
exports.ApplicationUpdateRequestNodeInfoRequestFailed = ApplicationUpdateRequestNodeInfoRequestFailed;
exports.ApplicationUpdateRequestNodeInfoRequestFailed = ApplicationUpdateRequestNodeInfoRequestFailed = __decorate([
    applicationUpdateType(ApplicationUpdateTypes.NodeInfo_RequestFailed)
], ApplicationUpdateRequestNodeInfoRequestFailed);
let ApplicationUpdateRequestNodeAdded = class ApplicationUpdateRequestNodeAdded extends ApplicationUpdateRequestWithNodeInfo {
};
exports.ApplicationUpdateRequestNodeAdded = ApplicationUpdateRequestNodeAdded;
exports.ApplicationUpdateRequestNodeAdded = ApplicationUpdateRequestNodeAdded = __decorate([
    applicationUpdateType(ApplicationUpdateTypes.Node_Added)
], ApplicationUpdateRequestNodeAdded);
let ApplicationUpdateRequestNodeRemoved = class ApplicationUpdateRequestNodeRemoved extends ApplicationUpdateRequest {
    constructor(host, options) {
        super(host, options);
        const { nodeId } = (0, core_1.parseNodeID)(this.payload, host.nodeIdType, 0);
        this.nodeId = nodeId;
        // byte 1/2 is 0, meaning unknown
    }
    nodeId;
};
exports.ApplicationUpdateRequestNodeRemoved = ApplicationUpdateRequestNodeRemoved;
exports.ApplicationUpdateRequestNodeRemoved = ApplicationUpdateRequestNodeRemoved = __decorate([
    applicationUpdateType(ApplicationUpdateTypes.Node_Removed)
], ApplicationUpdateRequestNodeRemoved);
let ApplicationUpdateRequestSmartStartHomeIDReceived = class ApplicationUpdateRequestSmartStartHomeIDReceived extends ApplicationUpdateRequest {
    constructor(host, options) {
        super(host, options);
        let offset = 0;
        const { nodeId, bytesRead: nodeIdBytes } = (0, core_1.parseNodeID)(this.payload, host.nodeIdType, offset);
        offset += nodeIdBytes;
        this.remoteNodeId = nodeId;
        // next byte is rxStatus
        offset++;
        this.nwiHomeId = this.payload.subarray(offset, offset + 4);
        offset += 4;
        const ccLength = this.payload[offset++];
        this.basicDeviceClass = this.payload[offset++];
        this.genericDeviceClass = this.payload[offset++];
        this.specificDeviceClass = this.payload[offset++];
        this.supportedCCs = (0, core_1.parseCCList)(this.payload.subarray(offset, offset + ccLength)).supportedCCs;
    }
    remoteNodeId;
    nwiHomeId;
    basicDeviceClass;
    genericDeviceClass;
    specificDeviceClass;
    supportedCCs;
    toLogEntry() {
        const message = {
            type: (0, shared_1.getEnumMemberName)(ApplicationUpdateTypes, this.updateType),
            "remote node ID": this.remoteNodeId,
            "NWI home ID": (0, shared_1.buffer2hex)(this.nwiHomeId),
            "basic device class": this.basicDeviceClass,
            "generic device class": this.genericDeviceClass,
            "specific device class": this.specificDeviceClass,
            "supported CCs": this.supportedCCs
                .map((cc) => `\n· ${(0, core_1.getCCName)(cc)}`)
                .join(""),
        };
        return {
            ...super.toLogEntry(),
            message,
        };
    }
};
exports.ApplicationUpdateRequestSmartStartHomeIDReceived = ApplicationUpdateRequestSmartStartHomeIDReceived;
exports.ApplicationUpdateRequestSmartStartHomeIDReceived = ApplicationUpdateRequestSmartStartHomeIDReceived = __decorate([
    applicationUpdateType(ApplicationUpdateTypes.SmartStart_HomeId_Received)
], ApplicationUpdateRequestSmartStartHomeIDReceived);
