// @zwave-js/server/dist/lib/driver/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DriverCommand = void 0;
var DriverCommand;
(function (DriverCommand) {
    DriverCommand["getConfig"] = "driver.get_config";
    DriverCommand["updateLogConfig"] = "driver.update_log_config";
    DriverCommand["getLogConfig"] = "driver.get_log_config";
    DriverCommand["enableStatistics"] = "driver.enable_statistics";
    DriverCommand["disableStatistics"] = "driver.disable_statistics";
    DriverCommand["isStatisticsEnabled"] = "driver.is_statistics_enabled";
    DriverCommand["startListeningLogs"] = "driver.start_listening_logs";
    DriverCommand["stopListeningLogs"] = "driver.stop_listening_logs";
    DriverCommand["checkForConfigUpdates"] = "driver.check_for_config_updates";
    DriverCommand["installConfigUpdate"] = "driver.install_config_update";
    DriverCommand["setPreferredScales"] = "driver.set_preferred_scales";
    DriverCommand["enableErrorReporting"] = "driver.enable_error_reporting";
    DriverCommand["softReset"] = "driver.soft_reset";
    DriverCommand["trySoftReset"] = "driver.try_soft_reset";
    DriverCommand["hardReset"] = "driver.hard_reset";
    DriverCommand["shutdown"] = "driver.shutdown";
    DriverCommand["updateOptions"] = "driver.update_options";
})(DriverCommand || (exports.DriverCommand = DriverCommand = {}));
