// zwave-js/build/lib/serialapi/network-mgmt/ReplaceFailedNodeRequest.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplaceFailedNodeRequestStatusReport = exports.ReplaceFailedNodeResponse = exports.ReplaceFailedNodeRequest = exports.ReplaceFailedNodeRequestBase = exports.ReplaceFailedNodeStatus = exports.ReplaceFailedNodeStartFlags = void 0;
const core_1 = require("@zwave-js/core");
const serial_1 = require("@zwave-js/serial");
var ReplaceFailedNodeStartFlags;
(function (ReplaceFailedNodeStartFlags) {
    ReplaceFailedNodeStartFlags[ReplaceFailedNodeStartFlags["OK"] = 0] = "OK";
    /** The replacing process was aborted because the controller  is not the primary one */
    ReplaceFailedNodeStartFlags[ReplaceFailedNodeStartFlags["NotPrimaryController"] = 2] = "NotPrimaryController";
    /** The replacing process was aborted because no call back function is used */
    ReplaceFailedNodeStartFlags[ReplaceFailedNodeStartFlags["NoCallbackFunction"] = 4] = "NoCallbackFunction";
    /** The replacing process aborted because the node was node found */
    ReplaceFailedNodeStartFlags[ReplaceFailedNodeStartFlags["NodeNotFound"] = 8] = "NodeNotFound";
    /** The replacing process is busy */
    ReplaceFailedNodeStartFlags[ReplaceFailedNodeStartFlags["ReplaceProcessBusy"] = 16] = "ReplaceProcessBusy";
    /** The replacing process could not be started*/
    ReplaceFailedNodeStartFlags[ReplaceFailedNodeStartFlags["ReplaceFailed"] = 32] = "ReplaceFailed";
})(ReplaceFailedNodeStartFlags || (exports.ReplaceFailedNodeStartFlags = ReplaceFailedNodeStartFlags = {}));
var ReplaceFailedNodeStatus;
(function (ReplaceFailedNodeStatus) {
    /* ZW_ReplaceFailedNode callback status definitions */
    ReplaceFailedNodeStatus[ReplaceFailedNodeStatus["NodeOK"] = 0] = "NodeOK";
    /** The failed node is ready to be replaced and controller is ready to add new node with the nodeID of the failed node. */
    ReplaceFailedNodeStatus[ReplaceFailedNodeStatus["FailedNodeReplace"] = 3] = "FailedNodeReplace";
    /** The failed node has been replaced. */
    ReplaceFailedNodeStatus[ReplaceFailedNodeStatus["FailedNodeReplaceDone"] = 4] = "FailedNodeReplaceDone";
    /** The failed node has not been replaced */
    ReplaceFailedNodeStatus[ReplaceFailedNodeStatus["FailedNodeReplaceFailed"] = 5] = "FailedNodeReplaceFailed";
})(ReplaceFailedNodeStatus || (exports.ReplaceFailedNodeStatus = ReplaceFailedNodeStatus = {}));
let ReplaceFailedNodeRequestBase = class ReplaceFailedNodeRequestBase extends serial_1.Message {
    constructor(host, options) {
        if ((0, serial_1.gotDeserializationOptions)(options)
            && new.target !== ReplaceFailedNodeRequestStatusReport) {
            return new ReplaceFailedNodeRequestStatusReport(host, options);
        }
        super(host, options);
    }
};
exports.ReplaceFailedNodeRequestBase = ReplaceFailedNodeRequestBase;
exports.ReplaceFailedNodeRequestBase = ReplaceFailedNodeRequestBase = __decorate([
    (0, serial_1.messageTypes)(serial_1.MessageType.Request, serial_1.FunctionType.ReplaceFailedNode),
    (0, serial_1.priority)(core_1.MessagePriority.Controller)
], ReplaceFailedNodeRequestBase);
let ReplaceFailedNodeRequest = class ReplaceFailedNodeRequest extends ReplaceFailedNodeRequestBase {
    constructor(host, options) {
        super(host, options);
        this.failedNodeId = options.failedNodeId;
    }
    // This must not be called nodeId or rejectAllTransactions may reject the request
    /** The node that should be removed */
    failedNodeId;
    serialize() {
        const nodeId = (0, core_1.encodeNodeID)(this.failedNodeId, this.host.nodeIdType);
        this.payload = Buffer.concat([nodeId, Buffer.from([this.callbackId])]);
        return super.serialize();
    }
};
exports.ReplaceFailedNodeRequest = ReplaceFailedNodeRequest;
exports.ReplaceFailedNodeRequest = ReplaceFailedNodeRequest = __decorate([
    (0, serial_1.expectedResponse)(serial_1.FunctionType.ReplaceFailedNode)
], ReplaceFailedNodeRequest);
let ReplaceFailedNodeResponse = class ReplaceFailedNodeResponse extends serial_1.Message {
    constructor(host, options) {
        super(host, options);
        this._replaceStatus = this.payload[0];
    }
    _replaceStatus;
    get replaceStatus() {
        return this._replaceStatus;
    }
    isOK() {
        return this._replaceStatus === ReplaceFailedNodeStartFlags.OK;
    }
};
exports.ReplaceFailedNodeResponse = ReplaceFailedNodeResponse;
exports.ReplaceFailedNodeResponse = ReplaceFailedNodeResponse = __decorate([
    (0, serial_1.messageTypes)(serial_1.MessageType.Response, serial_1.FunctionType.ReplaceFailedNode)
], ReplaceFailedNodeResponse);
class ReplaceFailedNodeRequestStatusReport extends ReplaceFailedNodeRequestBase {
    constructor(host, options) {
        super(host, options);
        this.callbackId = this.payload[0];
        this._replaceStatus = this.payload[1];
    }
    _replaceStatus;
    get replaceStatus() {
        return this._replaceStatus;
    }
    isOK() {
        return (this._replaceStatus
            === ReplaceFailedNodeStatus.FailedNodeReplaceDone);
    }
}
exports.ReplaceFailedNodeRequestStatusReport = ReplaceFailedNodeRequestStatusReport;
