// zwave-js/build/lib/node/_Types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zWaveNodeEvents = exports.NodeStatus = exports.InterviewStage = exports.ControllerStatus = exports.PowerlevelTestStatus = exports.Powerlevel = exports.MultilevelSwitchCommand = exports.FirmwareUpdateStatus = exports.EntryControlEventTypes = exports.EntryControlDataTypes = void 0;
var safe_1 = require("@zwave-js/cc/safe");
Object.defineProperty(exports, "EntryControlDataTypes", { enumerable: true, get: function () { return safe_1.EntryControlDataTypes; } });
Object.defineProperty(exports, "EntryControlEventTypes", { enumerable: true, get: function () { return safe_1.EntryControlEventTypes; } });
Object.defineProperty(exports, "FirmwareUpdateStatus", { enumerable: true, get: function () { return safe_1.FirmwareUpdateStatus; } });
Object.defineProperty(exports, "MultilevelSwitchCommand", { enumerable: true, get: function () { return safe_1.MultilevelSwitchCommand; } });
Object.defineProperty(exports, "Powerlevel", { enumerable: true, get: function () { return safe_1.Powerlevel; } });
Object.defineProperty(exports, "PowerlevelTestStatus", { enumerable: true, get: function () { return safe_1.PowerlevelTestStatus; } });
var safe_2 = require("@zwave-js/core/safe");
Object.defineProperty(exports, "ControllerStatus", { enumerable: true, get: function () { return safe_2.ControllerStatus; } });
Object.defineProperty(exports, "InterviewStage", { enumerable: true, get: function () { return safe_2.InterviewStage; } });
Object.defineProperty(exports, "NodeStatus", { enumerable: true, get: function () { return safe_2.NodeStatus; } });
exports.zWaveNodeEvents = [
    "notification",
    "interview failed",
    "firmware update progress",
    "firmware update finished",
    "wake up",
    "sleep",
    "dead",
    "alive",
    "interview completed",
    "ready",
    "interview stage completed",
    "interview started",
    "value added",
    "value updated",
    "value removed",
    "metadata updated",
    "value notification",
];
